src/nomad_simulations/schema_packages/data_types.py

```diff
@@ -0,0 +1,316 @@
+#
+# Copyright The NOMAD Authors.
+#
+# This file is part of NOMAD. See https://nomad-lab.eu for further info.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import re
+from collections.abc import Generator
+from typing import Any
+
+import numpy as np
+from nomad.metainfo.data_type import ExactNumber, InexactNumber
+
+# Match patterns like '[0,3)', '(0,5]', '[1,)', '(,10)', etc.
+bounds_patt = re.compile(r'^([\[\(])(-?\d*\.?\d*|),\s*(-?\d*\.?\d*|)([\]\)])$')
+
+
+def _flatten_values(data: Any) -> Generator[Any, None, None]:
+    """Generator that yields all scalar values from nested list/array structure."""
+    if isinstance(data, np.ndarray):
+        yield from data.flatten()
+    elif isinstance(data, list):
+        for item in data:
+            if isinstance(item, list):
+                yield from _flatten_values(item)
+            else:
+                yield item
+    else:
+        yield data
```

Comment:
Just return a list, this generator is consumed later as `list(_flatten_values(data))`, no point of using a generator.